### PR TITLE
[dockers] Update critical_processes file syntax

### PR DESCRIPTION
**Pull request by yozhao101**
Signed-off-by: Yong Zhao <yozhao@microsoft.com>

**- Why I did it**
Initially, the critical_processes file contains either the name of critical process or the name of group.
For example, the critical_processes file in the dhcp_relay container contains a single group name
`isc-dhcp-relay`. When testing the autorestart feature of each container, we need get all the critical
processes and test whether a  container can be restarted correctly if one of its critical processes is
killed. However, it will be difficult to differentiate whether the names in the critical_processes file are
the critical processes or group names. At the same time, changing the syntax in this file will separate the individual process from the groups and also makes it clear to the user.

Right now the critical_processes file contains two different kind of entries. One is "program:xxx" which indicates a critical process. Another is "group:xxx" which indicates a group of critical processes
managed by supervisord using the name "xxx". At the same time, I also updated the logic to
parse the file critical_processes in supervisor-proc-event-listener script.

**- How I did it**

**- How to verify it**
We can first enable the autorestart feature of a specified container for example `dhcp_relay` by running the comman `sudo config container feature autorestart dhcp_relay enabled` on DUT. Then we can select a critical process from the command `docker top dhcp_relay` and use the command `sudo kill -SIGKILL <pid>` to kill that critical process. Final step is to check whether the container is restarted correctly or not.

**- Description for the changelog**

### CODE REVIEW

File: files/scripts/supervisor-proc-exit-listener [hunk 1]
```diff
@@ -10,14 +10,28 @@ import swsssdk
 
 from supervisor import childutils
 
-# Contents of file should be the names of critical processes (as defined in
-# supervisor.conf file), one per line
+# Contents of file should be the names of critical processes and group name
```
Comment (jleveque) [inline files/scripts/supervisor-proc-exit-listener]: Reword this comment as follows:

```
Each line of this file should specify either one critical process or one critical process group, (as defined in supervisor.conf file), in the following format:

program:<program_name>
group:<group_name>
```

File: files/scripts/supervisor-proc-exit-listener [hunk 2]
```diff
@@ -10,14 +10,28 @@ import swsssdk
...
-# supervisor.conf file), one per line
+# Contents of file should be the names of critical processes and group name 
+# (as defined in supervisor.conf file), one per line
 CRITICAL_PROCESSES_FILE = '/etc/supervisor/critical_processes'
 
 # This table in databse contains the features for container and each
 # feature for a row will be configured a state or number.
 CONTAINER_FEATURE_TABLE_NAME = 'CONTAINER_FEATURE'
 
+# Read the list of critical processes and group names from a file
```
Comment (jleveque) [inline files/scripts/supervisor-proc-exit-listener:21]: Reword:

```
# Read the critical process/group names from CRITICAL_PROCESSES_FILE 
```

File: files/scripts/supervisor-proc-exit-listener [hunk 3]
```diff
@@ -10,14 +10,28 @@ import swsssdk
...
+def get_group_list_and_critical_processes():
+    group_list = []
+    critical_processes = [] 
+
+    with open(CRITICAL_PROCESSES_FILE, 'r') as file:
+	for line in file:
+            if line.find("group") != -1:
```
Comment (jleveque) [inline files/scripts/supervisor-proc-exit-listener:28]: `find()` is not strict enough. What if the process or group name contains "group"?

You should split the line by the `:` character and check if the first token is "group" or "program" (log an error if neither).

File: files/scripts/supervisor-proc-exit-listener [hunk 4]
```diff
@@ -10,14 +10,28 @@ import swsssdk
...
 CRITICAL_PROCESSES_FILE = '/etc/supervisor/critical_processes'
 
 # This table in databse contains the features for container and each
 # feature for a row will be configured a state or number.
 CONTAINER_FEATURE_TABLE_NAME = 'CONTAINER_FEATURE'
 
+# Read the list of critical processes and group names from a file
+def get_group_list_and_critical_processes():
+    group_list = []
+    critical_processes = []
```
Comment (jleveque) [inline files/scripts/supervisor-proc-exit-listener:24]: Prefer consistent naming. Either `critical_group_list` and `critical_process_list` *or* `critical_groups` and `critical_processes`

Comment (yozhao101) [reply files/scripts/supervisor-proc-exit-listener]: Updated.

File: files/scripts/supervisor-proc-exit-listener (hunk 3 above)
Comment (yozhao101) [reply files/scripts/supervisor-proc-exit-listener]: Great suggestion! updated.

File: files/scripts/supervisor-proc-exit-listener (hunk 2 above)
Comment (yozhao101) [reply files/scripts/supervisor-proc-exit-listener]: Updated!

File: files/scripts/supervisor-proc-exit-listener (hunk 1 above)
Comment (yozhao101) [reply files/scripts/supervisor-proc-exit-listener]: Updated!

**Comment by yozhao101** [comment]
> Please fix conflicts. Looks like your repo was out-of-date. I recently renamed "docker-lldp-sv2" to "docker-lldp" and "docker-snmp-sv2" to "docker-snmp".

Fixed the conflicts.

**Comment by jleveque** [comment]
@yozhao101: This will not cherry-pick cleanly to the 201911 branch due to the directory name changes and the absence of new containers, so you will also need to open a separate PR against that branch.

**Comment by yozhao101** [comment]
> @yozhao101: This will not cherry-pick cleanly to the 201911 branch due to the directory name changes and the absence of new containers, so you will also need to open a separate PR against that branch.

Yes, I will open a new PR against 201911 branch.

**Comment by lguohan** [comment]
retest broadcom please

**Comment by jleveque** [comment]
Retest broadcom please

**Comment by jleveque** [comment]
Retest broadcom please

**Pull request merged**